### ISSUE TITLE
1D version

### ISSUE THREAD

**Issue by floboc** [stfc-sciml/DifferentialConv2d#1]
Hello,
Thanks a lot for providing the code for your paper.
Would it be easy to adapt this for 1D convolutions?
Thanks!

**Comment by floboc** [comment]
Maybe actually just using Conv2D with H=1 is sufficient, but I assume you only provide square kernels?

**Comment by kuangdai** [comment]
No, currently the code is prepared for 3x3, 5x5, and 7x7. Let me evaluate how to do this quickly. What is the kernel size for your 1D application?

**Comment by floboc** [comment]
That would be awesome, thanks a lot!
I have kernel size of both 3 and 5

**Comment by floboc** [comment]
Another question that might be related to this: in my case I have to deal with varying length data, so each sample in my batch has a different length, and I have an additional input specifying a mask to indicate valid samples.

Like if my batch has a sequence length of 100 and the 5 sample has an original length of 75, then the 5th row of the mask will have 75 ones followed by zeros.

Can your method be easily adapted to take into account such a mask and not only handle the padding at the borders of the batch?

**Comment by OlliNiemitalo** [comment]
In case it helps, I have source code here to generate transformation matrixes for any kernel size: [gen_constants.py](https://github.com/hamk-uas/DifferentialConv2d/blob/main/diff_conv2d/lagrange_constants/gen_constants.py).

Looking at the transformation matrixes, I think that the transformation can be implemented in a horizontal followed by a vertical transformation step, using 1-d Lagrange polynomial extrapolation coefficients, see [lagrange_padding_coefs.json](https://github.com/hamk-uas/DifferentialConv2d/blob/main/diff_conv2d/lagrange_constants/lagrange_padding_coefs.json), for the taps that end up outside the window.

**Comment by kuangdai** [comment]
First, after reviewing the theory, I can confirm that, for the 1D case (but not for the 2D case as presented in the paper), our method is equivalent to padding the 1D array by Lagrange extrapolation then do conv1d with no padding. 

Based on this fact, I made the attached code to do what you want (hopefully). You use the two functions for padding then do conv1d with no padding. These functions are differentiable, so you can use them in training.

Because you mentioned variable length for each sample, I do left and right padding separately, assuming that your data are aligned on the left with masks (nan is used in the code) on the right.


```python
import torch

lagrange_padding_coefficients = {
    3: [[1.0, -3.0, 3.0]],
    5: [[1.0, -5.0, 10.0, -10.0, 5.0],
        [5.0, -24.0, 45.0, -40.0, 15.0]]
}


def pad_lagrange_left(x, kernel_size=3):
    """
    Padding by Lagrange extrapolation on the left
    :param x: input tensor with shape [batch_size, length]
    :param kernel_size: kernel size
    :return: padded tensor
    """
    # get coefficients
    coeff = torch.tensor(lagrange_padding_coefficients[kernel_size],
                         dtype=x.dtype, device=x.device).flip(dims=[0, 1])
    n, p = coeff.shape
    # create output
    out = torch.full((x.shape[0], x.shape[1] + n), torch.nan,
                     dtype=x.dtype, device=x.device)
    # copy old
    out[:, n:] = x
    # pad new
    out[:, :n] = torch.einsum('np,bp->bn', coeff, x[:, :p])
    return out


def pad_lagrange_right(x, kernel_size=3, sample_length=None):
    """
    Padding by Lagrange extrapolation on the right
    :param x: input tensor with shape [batch_size, length]
    :param kernel_size: kernel size
    :param sample_length: length of each sample, with shape [batch_size]
    :return: padded tensor
    """
    # use pad left if no sample-wise length
    if sample_length is None:
        return pad_lagrange_left(
            x.flip(dims=[1]), kernel_size=kernel_size).flip(dims=[1])

    # get coefficients
    coeff = torch.tensor(lagrange_padding_coefficients[kernel_size],
                         dtype=x.dtype, device=x.device).flip(dims=[1])
    n, p = coeff.shape
    # create output
    out = torch.full((x.shape[0], x.shape[1] + n), torch.nan,
                     dtype=x.dtype, device=x.device)
    # copy old
    out[:, :-n] = x
    # pad new
    batch_index = torch.arange(x.shape[0]).unsqueeze(-1)
    out_length_index = torch.tensor(sample_length)[:, None] + torch.arange(0, n)[None, :]
    x_length_index = torch.tensor(sample_length)[:, None] - torch.arange(0, p)[None, :] - 1
    out[batch_index, out_length_index] = torch.einsum('np,bp->bn', coeff,
                                                      x[batch_index, x_length_index])
    return out


if __name__ == "__main__":
    x_ = torch.tensor([[0, 1, 2, 3, 4, 5, 6],
                       [2, 3, 4, 5, 6, 7, torch.nan]], dtype=torch.float)
    print(pad_lagrange_left(x_, kernel_size=5))
    print(pad_lagrange_right(x_, kernel_size=5, sample_length=[7, 6]))
```

**Comment by floboc** [comment]
Thanks a lot for your fast answer!
I think teh code will need a bit of adaptation as it seems to assume no channel dimension (ie input of shape BxL instead of BxCxL).

**Comment by kuangdai** [comment]
At the step of padding, B and C are both batch dimensions, as padding only operates on L. You can simply reshape input from [B, C, L] to [BxC, L] and reshape output back to [B, C, L + n].

**Comment by floboc** [comment]
ah yes simple indeed.
Thanks a lot for your help, much appreciated!

**Comment by floboc** [comment]
Me again, I think there might a small issue with the above code, as it injects more NaNs in the output than there are in the input:

If I replace your sample code to perform the entire process (padding + conv):
```
if __name__ == "__main__":
    x_ = torch.tensor([[0, 1, 2, 3, 4, 5, 6],
                       [2, 3, 4, 5, 6, 7, torch.nan]], dtype=torch.float)
    print(pad_lagrange_left(x_, kernel_size=5))
    print(pad_lagrange_right(x_, kernel_size=5, sample_length=[7, 6]))

    x_padded = pad_lagrange_right(pad_lagrange_left(x_, kernel_size=5), kernel_size=5, sample_length=[7, 6])
    print(x_padded)

    conv = torch.nn.Conv1d(1, 1, kernel_size=5, padding=0)
    y = conv(x_padded.unsqueeze(1))
    print(y)
```

Then the output is something like:
```
tensor([[[0.0475, 0.4631, 0.8787, 1.2944, 1.7100,    nan,    nan]],

        [[0.8787, 1.2944, 1.7100, 2.1257,    nan,    nan,    nan]]],
```

**Comment by kuangdai** [comment]
That’s because, after padding left, your sample_length should be [9, 8] instead of [7, 6]. Or you can do right padding before

On 25 Jan 2024, at 15:38, Florent Bocquelet ***@***.***> wrote:



Me again, I think there might a small issue with the above code, as it injects more NaNs in the output than there are in the input:

If I replace your sample code to perform the entire process (padding + conv):

if __name__ == "__main__":
    x_ = torch.tensor([[0, 1, 2, 3, 4, 5, 6],
                       [2, 3, 4, 5, 6, 7, torch.nan]], dtype=torch.float)
    print(pad_lagrange_left(x_, kernel_size=5))
    print(pad_lagrange_right(x_, kernel_size=5, sample_length=[7, 6]))

    x_padded = pad_lagrange_right(pad_lagrange_left(x_, kernel_size=5), kernel_size=5, sample_length=[7, 6])
    print(x_padded)

    conv = torch.nn.Conv1d(1, 1, kernel_size=5, padding=0)
    y = conv(x_padded.unsqueeze(1))
    print(y)


Then the output is something like:

tensor([[[0.0475, 0.4631, 0.8787, 1.2944, 1.7100,    nan,    nan]],

        [[0.8787, 1.2944, 1.7100, 2.1257,    nan,    nan,    nan]]],

**Comment by floboc** [comment]
Yeah I just realized that myself almost immediately which is why I deleted my previous message but it seems you still got it!
Thanks for answering!
